fix: Support for msg[content] as a list

KrishnanPrash opened the pull request:
#### Overview:


Chat templates have conflicting expectations for message content format:
- **Standard templates**  expect strings: `"Hello"`
- **Multimodal templates** (llava) expect arrays: `[{"type": "text", "text": "Hello"}]`

When the wrong format is provided, content goes missing or renders as malformed in prompts.

1. **Detection at model load:** Test-render the template with both formats to detect requirements
2. **Normalization per request:** Convert between formats based on template needs
   - Standard templates: Flatten text-only arrays → strings (`"t1\nt2"`)
   - Multimodal templates: Wrap strings → arrays (`[{"type": "text", "text": "..."}]`)
3. **Smart preservation:** Mixed content (text + images) always kept as-is


#### Details:

- Added `detect_content_array_usage()` in `formatters.rs`
- Added `requires_content_arrays` field to `HfTokenizerConfigJsonFormatter` 
- Made `may_be_fix_msg_content()` bidirectional with `preserve_arrays` parameter
- Updated render pipeline to apply normalization automatically

#### Related PRs:

- #3067
- #4379 

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->

## Summary by CodeRabbit

* **Improvements**
  * Enhanced message content formatting to better detect and handle different template styles.
  * Improved flexibility in converting between different message content formats for better template compatibility.

<!-- end of auto-generated comment: release notes by coderabbit.ai -->

keivenchang committed:
feat: support multimodal models with non-standard Jinja2 tags

Implement placeholder replacement for non-standard Jinja2 tags (e.g., {% generation %})
that minijinja doesn't recognize but vLLM's custom extensions require.

Rust frontend replaces tags with __JINJA_BLOCK_<TAG>__ for validation, Python backend
restores them before vLLM processing. Also fix model card parsing for multimodal models:
make num_hidden_layers optional and handle eos_token_id as single value or array.

Signed-off-by: Keiven Chang <keivenchang@users.noreply.github.com>

KrishnanPrash committed:
Handle chat templates that expect message[content] be a string

Signed-off-by: Krishnan Prashanth <kprashanth@nvidia.com>

KrishnanPrash committed:
Cleaning up comments + variable names + structure

Signed-off-by: Krishnan Prashanth <kprashanth@nvidia.com>

KrishnanPrash committed:
fix: cargo_fmt fix

Signed-off-by: Krishnan Prashanth <kprashanth@nvidia.com>

rmccorm4 commented on lib/llm/src/preprocessor/prompt/template/formatters.rs:
```diff
@@ -165,11 +201,20 @@ impl HfTokenizerConfigJsonFormatter {
         }
 
+        // Detect at model load time whether this template requires content arrays
+        let requires_content_arrays = detect_content_array_usage(&env);
```
How does vllm-serve, sglang, trtllm-serve, etc. solve this problem?

KrishnanPrash replied:
From previous investigations, vLLM's pre-processor does something like:
-  Feeds the chat template to Jinja2, which internally parses and generates an AST.
-  Relies on internal representations of Jinja2 to check if a for loop is done over `messages[content]`.
- If yes, leave `messages[content]` as an array.
- If no, flatten `messages[content]` to a string. (Similar to what we do in #3067)

And from my limited investigation into `MiniJinja` it does not expose anything similar to that.

keivenchang reviewed (approved):
I was really sad to learn from Prashanth that the JSON formats are not versioned (or, specify versions), and some model makers (Mistral) don't even publish formal specs. So, I guess the best thing to do then is to keep writing code to detect changes in JSON.

Having that said, the templates will just keep changing every once in a while-- even many times throughout the year. It would be nice if we (Dynamo) document all the different types of templates that we currently can handle (and future ones that we are working on). It'll help programmers understand the different formats, as well as get picked up by Cursor/Claude to write code to support these examples. Can you add something like this in a *.md file? It'll help you (and AI) tremendously.

keivenchang committed:
fix: Handle two non Jinja2 tags in chat templates

- Add remove_known_non_jinja2_tags to strip {% generation %} tags before minijinja validation
- Fixes LLaVA and other models using vLLM-specific template extensions
- Make num_hidden_layers optional in model card parsing for multimodal compatibility
- Handle eos_token_id as single value or array in generation_config.json

Signed-off-by: Keiven Chang <keivenchang@users.noreply.github.com>

KrishnanPrash committed:
Merge branch 'keivenchang/MDC-fix-on-main-nvbugs5662072' into kprashanth/msg_content_list_or_string

Signed-off-by: KrishnanPrash <140860868+KrishnanPrash@users.noreply.github.com>

KrishnanPrash committed:
Fixing unintended changes

Signed-off-by: Krishnan Prashanth <kprashanth@nvidia.com>

KrishnanPrash committed:
Adding Llava test

Signed-off-by: Krishnan Prashanth <kprashanth@nvidia.com>

KrishnanPrash committed:
Merge branch 'main' into kprashanth/msg_content_list_or_string

Signed-off-by: KrishnanPrash <140860868+KrishnanPrash@users.noreply.github.com>
